File: 03/gpt-5.1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var totalJoltagePart1 int64
	var totalJoltagePart2 int64

	for scanner.Scan() {
		line := scanner.Text()
		if len(line) == 0 {
			continue
		}

		maxJoltagePart1 := 0
		found := false

		for d1 := 9; d1 >= 1 && !found; d1-- {
			c1 := byte('0' + d1)
			idx1 := indexByte(line, c1)
			if idx1 < 0 {
				continue
			}

			for d2 := 9; d2 >= 1; d2-- {
				c2 := byte('0' + d2)
				idx2 := lastIndexByte(line, c2)
				if idx2 < 0 {
					continue
				}

				if idx1 < idx2 {
					maxJoltagePart1 = d1*10 + d2
					found = true
					break
				}
			}
		}

		totalJoltagePart1 += int64(maxJoltagePart1)

		// Part 2: choose exactly 12 digits to form the largest possible number
		digits := line
		targetLength := 12
		chosen := make([]byte, targetLength)
		start := 0

		for pos := 0; pos < targetLength; pos++ {
			remainingNeeded := targetLength - pos - 1
			bestDigit := byte('0')
			bestIndex := -1

			for d := 9; d >= 0; d-- {
				ch := byte('0' + d)
				lastIndex := lastIndexByteFrom(digits, ch, start)
				if lastIndex < start {
					continue
				}

				if lastIndex <= len(digits)-1-remainingNeeded {
					bestDigit = ch
					bestIndex = lastIndex
					break
				}
			}

			if bestIndex == -1 {
				bestDigit = digits[start]
				bestIndex = start
			}

			chosen[pos] = bestDigit
			start = bestIndex + 1
		}

		lineJoltagePart2 := int64(0)
		for _, ch := range chosen {
			lineJoltagePart2 = lineJoltagePart2*10 + int64(ch-'0')
		}

		totalJoltagePart2 += lineJoltagePart2
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	fmt.Printf("Day 03 Part 1: %d\n", totalJoltagePart1)
	fmt.Printf("Day 03 Part 2: %d\n", totalJoltagePart2)
}

func indexByte(s string, c byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == c {
			return i
		}
	}
	return -1
}

func lastIndexByte(s string, c byte) int {
	for i := len(s) - 1; i >= 0; i-- {
		if s[i] == c {
			return i
		}
	}
	return -1
}

func lastIndexByteFrom(s string, c byte, start int) int {
	last := -1
	for i := start; i < len(s); i++ {
		if s[i] == c {
			last = i
		}
	}
	return last
}

File: 03/gpt-5.1/rust/src/main.rs
use std::fs;

fn main() {
    let contents = fs::read_to_string("input.txt").expect("Failed to read input");

    let mut total_joltage_part1: i64 = 0;
    let mut total_joltage_part2: i64 = 0;

    for line in contents.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let mut max_joltage_part1 = 0;
        let mut found = false;

        for d1 in (1..=9).rev() {
            if found {
                break;
            }

            let c1 = (b'0' + d1 as u8) as char;
            let idx1 = line.find(c1);
            if idx1.is_none() {
                continue;
            }
            let idx1 = idx1.unwrap();

            for d2 in (1..=9).rev() {
                let c2 = (b'0' + d2 as u8) as char;
                let idx2 = line.rfind(c2);
                if idx2.is_none() {
                    continue;
                }
                let idx2 = idx2.unwrap();

                if idx1 < idx2 {
                    max_joltage_part1 = d1 * 10 + d2;
                    found = true;
                    break;
                }
            }
        }

        total_joltage_part1 += max_joltage_part1;

        // Part 2: choose exactly 12 digits to form the largest possible number in order
        let digits = line.as_bytes();
        let target_length = 12usize;
        let mut chosen: Vec<u8> = Vec::with_capacity(target_length);
        let mut start = 0usize;

        for pos in 0..target_length {
            let remaining_needed = target_length - pos - 1;
            let mut best_digit: Option<u8> = None;
            let mut best_index: Option<usize> = None;

            for d in (0u8..=9u8).rev() {
                let ch = b'0' + d;
                let mut last_index: Option<usize> = None;
                for i in start..digits.len() {
                    if digits[i] == ch {
                        last_index = Some(i);
                    }
                }

                if let Some(idx) = last_index {
                    if idx <= digits.len() - 1 - remaining_needed {
                        best_digit = Some(ch);
                        best_index = Some(idx);
                        break;
                    }
                }
            }

            let (chosen_digit, idx) = if let (Some(d), Some(i)) = (best_digit, best_index) {
                (d, i)
            } else {
                (digits[start], start)
            };

            chosen.push(chosen_digit);
            start = idx + 1;
        }

        let mut line_joltage_part2: i64 = 0;
        for ch in chosen {
            line_joltage_part2 = line_joltage_part2 * 10 + (ch - b'0') as i64;
        }

        total_joltage_part2 += line_joltage_part2;
    }

    println!("Day 03 Part 1: {}", total_joltage_part1);
    println!("Day 03 Part 2: {}", total_joltage_part2);
}
